NeuroID/src/main/java/com/neuroid/tracker/service/NIDServiceTracker.kt

```diff
@@ -39,91 +42,86 @@ object NIDServiceTracker {
 
-            val listJson = listEvents.map {
-                if (it.contains("\"CREATE_SESSION\"")) {
-                    JSONObject(
-                        it.replace(
-                            "\"url\":\"\"",
-                            "\"url\":\"$ANDROID_URI$firstScreenName\""
-                        )
-                    )
-                } else {
-                    JSONObject(it)
-                }
+        try {
+            val listEvents = getEvents(context, events)
+
+            if (listEvents.isEmpty()) {
+                // nothing to send
+                return Pair(NID_ERROR_SERVICE, false)
             }
 
-            val jsonListEvents = JSONArray(listJson)
+            stopLoopService = listEvents.last().get("type") == USER_INACTIVE
 
-            val data = getContentJson(context, jsonListEvents)
+            data = getContentJson(context, JSONArray(listEvents))
                 .replace("\\/", "/")
-            val stopLoopService = listEvents.last().contains(USER_INACTIVE)
-            NIDLog.d("NeuroID", "payload Json::: $data")
-
-            var retryAttempts = 0
-            while (retryAttempts < MAX_RETRY_ATTEMPTS) {
-                try {
-                    val os: OutputStream = conn.outputStream
-                    val writer = BufferedWriter(OutputStreamWriter(os, "UTF-8"))
-                    writer.write(data)
-                    writer.flush()
-                    writer.close()
-                    os.close()
-
-                    val code = conn.responseCode
-                    val message = conn.responseMessage
-
-                    if (code == 200) {
-                        NIDLog.d("NeuroID", "Http response code: $code")
-                        return Pair(code, stopLoopService)
-                    } else {
-                        NIDLog.e("NeuroID", "Error service: $message Code:$code")
-                        // Retry in case of HTTP error
-                        retryAttempts++
-                    }
-                } catch (ex: Exception) {
-                    NIDLog.e("NeuroID", "An error has occurred: ${ex.message}")
+
+            NIDLog.d("NeuroID", "payload: ${listEvents.size} events; ${data.length} bytes")
+
+            NeuroID.getInstance()?.saveIntegrationHealthEvents()
+        } catch (exception: OutOfMemoryError) {
+            // make a best effort attempt to continue and send an out of memory event
+            data = oomPayload
+        }
+
+        // Allow for override of this URL in config
+        NIDLog.d("NeuroID", "Url: $endpoint")
+        val url = URL("${endpoint}/${key}")
+
+        for (retryAttempts in 0 until MAX_RETRY_ATTEMPTS) {
+            val conn: HttpURLConnection = url.openConnection() as HttpURLConnection
+            try {
+                conn.requestMethod = "POST"
+                conn.doInput = true
+                conn.doOutput = true
+                conn.readTimeout = 5000
+                conn.connectTimeout = 5000
```

Comment:
Did you want to adjust these timeouts?

Reply:
No, just moved the connection configuration inside of the retry loop (which it could be argued that it should be left alone).